test/apps/iodemo/io_demo.cc

```diff
@@ -164,6 +165,141 @@ class P2pDemoCommon : public UcxContext {
         XFER_TYPE_RECV
     } xfer_type_t;
 
+    class BufferChunk {
+    public:
+        BufferChunk(size_t size, MemoryPool<BufferChunk> *pool) :
+            _capacity(size), _buffer(memalign(ALIGNMENT, size)), _size(0),
+            _pool(*pool) {
+            if (_buffer == NULL) {
+                throw std::bad_alloc();
+            }
+        }
+
+        void release() {
+            _pool.put(this);
+        }
+
+        inline void *buffer(size_t offset = 0) const {
+            return (uint8_t*)_buffer + offset;
+        }
+
+        inline void resize(size_t size) {
+            assert(size <= _capacity);
+            _size = size;
+        }
+
+        inline size_t size() const {
+            return _size;
+        }
+
+        ~BufferChunk() {
+            free(_buffer);
+        }
+
+    public:
+        const size_t              _capacity;
+
+    private:
+        void *                    _buffer;
+        size_t                    _size;
+        MemoryPool<BufferChunk> & _pool;
+    };
+
+    class BufferIov {
+    public:
+        BufferIov(size_t size, MemoryPool<BufferIov> *pool) : _pool(*pool) {
+            _iov.reserve(size);
+        }
+
+        size_t size() const {
+            return _iov.size();
+        }
+
+        void init(size_t data_size, MemoryPool<BufferChunk> &chunk_pool,
+                  uint32_t sn, bool fill) {
+            assert(_iov.empty());
+
+            BufferChunk *chunk = chunk_pool.get();
+            _iov.resize(get_chunk_cnt(data_size, chunk->_capacity));
```

Comment:
could be malloc on fast path?

Reply:
nope, ctor does `reserve()`

Reply:
and vector will not try to realloc or release mem after that?

Reply:
yes, but only if it's reserved correctly and capacity is not increased by resize()
(need to fix ctor param on initialization)